File: python/valid_palindrome_test.py
import unittest
from dataclasses import dataclass

from valid_palindrome import Solution


class TestSolution(unittest.TestCase):
    def cases(self):
        @dataclass
        class Case:
            given: str
            expected: bool

        return [
            Case(given=" ", expected=True),
            Case(given="abba", expected=True),
            Case(given="Ab,:c cba", expected=True),
            Case(given="101101", expected=True),
            Case(given="A man, a plan, a canal: Panama", expected=True),
            Case(given="race a car", expected=False),
        ]

    def test_is_palindrome(self):
        for case in self.cases():
            sol = Solution()
            assert case.expected == sol.isPalindrome(case.given)

File: go/valid_palindrome/valid_palindrome_test.go
package kata

import (
  "testing"

  "github.com/stretchr/testify/assert"
)

func TestIsPalindrome(t *testing.T) {
  testCases := []struct {
    name     string
    input    string
    expected bool
  }{
    // Add your test cases here
    {"empty string", "", true},
    {"single character", "a", true},
    {"two same characters", "aa", true},
    {"two different characters", "ab", false},
    {"palindrome even length", "abba", true},
    {"palindrome odd length", "racecar", true},
    {"not a palindrome", "hello", false},
    {"palindrome with spaces", "A man a plan a canal Panama", true},
    {"palindrome with punctuation", "Madam, I'm Adam.", true},
    {"not a palindrome with punctuation", "Hello, World!", false},
  }

  for _, tc := range testCases {
    t.Run(tc.name, func(t *testing.T) {
      result := isPalindrome(tc.input)
      assert.Equal(t, tc.expected, result)
    })
  }
}
